# **(EDA)**

## Objectives

* Write your notebook objective here, for example, "Fetch data from Kaggle and save as raw data", or "engineer features for modelling"

## Inputs

* Write down which data or information you need to run the notebook 

## Outputs

* Write here which files, code or artefacts you generate by the end of the notebook 

## Additional Comments

* If you have any additional comments that don't fit in the previous bullets, please state them here. 



---

# Change working directory

* We are assuming you will store the notebooks in a subfolder, therefore when running the notebook in the editor, you will need to change the working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [3]:
import os
current_dir = os.getcwd()
current_dir

'c:\\Users\\Ewa\\Documents\\vscode-projects\\Rest_Wellness_and_Lifestyle\\02_jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [4]:
os.chdir(os.path.dirname(current_dir))
print("Rest_Wellness_and_Lifestyle")

Rest_Wellness_and_Lifestyle


Confirm the new current directory

In [5]:
current_dir = os.getcwd()
current_dir

'c:\\Users\\Ewa\\Documents\\vscode-projects\\Rest_Wellness_and_Lifestyle'

# Section 1

Import python's libraries

In [26]:
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.io as pio
pio.renderers.default = "notebook"
import plotly.graph_objects as go
import plotly.subplots as sp
from scipy.stats import gaussian_kde


---

# Section 2

Load clean dataset

In [14]:
# Load cleaned dataset
df = pd.read_csv("/Users/Ewa/Documents/vscode-projects/Rest_Wellness_and_Lifestyle/01_data/Sleep_health_and_lifestyle_cleaned_dataset.csv")


In [15]:
# Display basic information about the dataset
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 369 entries, 0 to 368
Data columns (total 15 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Gender                   369 non-null    object 
 1   Age                      369 non-null    int64  
 2   Occupation               369 non-null    object 
 3   Sleep Duration           369 non-null    float64
 4   Quality of Sleep         369 non-null    int64  
 5   Physical Activity Level  369 non-null    int64  
 6   Stress Level             369 non-null    int64  
 7   BMI Category             369 non-null    object 
 8   Blood Pressure           369 non-null    object 
 9   Heart Rate               369 non-null    int64  
 10  Daily Steps              369 non-null    int64  
 11  Sleep Disorder           153 non-null    object 
 12  Systolic Pressure        369 non-null    int64  
 13  Diastolic Pressure       369 non-null    int64  
 14  Blood Pressure Category  3

Gender  Age   Occupation  Sleep Duration  Quality of Sleep  \
0   Male   27     Engineer             6.1                 6   
1   Male   28       Doctor             6.2                 6   
2   Male   28       Doctor             6.2                 6   
3   Male   28  Salesperson             5.9                 4   
4   Male   28  Salesperson             5.9                 4   

   Physical Activity Level  Stress Level BMI Category Blood Pressure  \
0                       42             6   Overweight         126/83   
1                       60             8       Normal         125/80   
2                       60             8       Normal         125/80   
3                       30             8        Obese         140/90   
4                       30             8        Obese         140/90   

   Heart Rate  Daily Steps Sleep Disorder  Systolic Pressure  \
0          77         4200            NaN                126   
1          75        10000            NaN                125   
2          75        10000            NaN                125   
3          85         3000    Sleep Apnea                140   
4          85         3000    Sleep Apnea                140   

   Diastolic Pressure Blood Pressure Category  
0                  83                  Normal  
1                  80                  Normal  
2                  80                  Normal  
3                  90            Hypertension  
4                  90            Hypertension

---

Initial Observations:

The dataset contains 369 rows and 16 columns.
It includes a mix of numeric (e.g., Age, Sleep Duration, Heart Rate) and categorical (e.g., Gender, BMI Category, Sleep Disorder) variables.
No missing values are present
There are some redundant columns (e.g., "Blood Pressure" and its split values "Systolic Pressure" & "Diastolic Pressure").

In [16]:
#Get categorical statistics
categorical_stats = df.describe(include='object')
display(categorical_stats)

Gender Occupation BMI Category Blood Pressure Sleep Disorder  \
count     369        369          369            369            153   
unique      2          7            3             24              2   
top      Male      Nurse       Normal         130/85       Insomnia   
freq      189         73          216             99             77   

       Blood Pressure Category  
count                      369  
unique                       3  
top                     Normal  
freq                       228

In [17]:
# Describe basic statistics of the dataset

summary_stats = df.describe()
styled_summary = summary_stats.style.background_gradient(cmap='Blues')
display(styled_summary)

NOTE

Dataset Summary:

* Age:

Mean (42.27) / Median (43): The dataset reflects a somewhat older adult population . The close values of the mean and median suggest the age distribution is fairly symmetrical.
Standard Deviation (8.68): The age distribution has a moderate spread, with a mix of younger and older individuals around the average.
Range (27 - 59): The dataset captures a wide age range of adults.

* Sleep Duration:

Mean (7.14) / Median (7.20): On average, individuals in the dataset get about 7 hours of sleep per day. The proximity of the mean and median suggests a relatively symmetrical distribution.
Range (5.9 - 8.5): There’s some variation in sleep duration, but most individuals sleep within a fairly narrow range.

* Quality of Sleep:

Mean (7.33) / Median (7): The average sleep quality rating is above the midpoint of the 1-10 scale. The slightly higher mean compared to the median suggests the distribution may be slightly right-skewed (positive skew), meaning that while most individuals report average to good sleep quality, a few individuals with significantly higher sleep quality ratings are pulling the mean up.

* Physical Activity Level:

Mean (59.37) / Median (60): On average, individuals engage in around an hour of physical activity per day.
Range (30 - 90): There is considerable variation in activity levels, with some individuals being significantly more active than others.

* Stress Level:

Mean (5.36) / Median (5): The average stress level is slightly above the midpoint of the 1-10 scale. The lower mean compared to the median suggests a possible skew towards higher stress levels for certain individuals.
 
 * Heart Rate:

Mean (70.06) / Median (70): The average resting heart rate is approximately 70 beats per minute, within the normal range for adults. The near-identical mean and median suggest a balanced distribution.

* Daily Steps:

Mean (6836.31) / Median (7000): On average, individuals take nearly 7,000 steps per day.
Range (3000 - 10000): There’s considerable variation in daily step counts, indicating varying levels of activity among individuals.

In [ ]:
# Data distribution of numerical features (histograms with trend lines) - code created using Chat GPT 
# Selecting numerical columns
numerical_columns = df.select_dtypes(include=['number']).columns

# Creating subplots with improved layout and styling
fig = sp.make_subplots(
    rows=len(numerical_columns), cols=1, 
    subplot_titles=[f'Distribution of {col}' for col in numerical_columns],
    vertical_spacing=0.08
)

# Colorblind-friendly palette
colorblind_palette = ["#377eb8", "#ff7f00", "#4daf4a", "#f781bf", "#a65628", "#984ea3", "#999999"]

# Adding histograms and trend lines
for i, column in enumerate(numerical_columns):
    hist_data = df[column].dropna()
    color = colorblind_palette[i % len(colorblind_palette)]  # Cycle through the palette
    
    # Histogram trace
    fig.add_trace(
        go.Histogram(
            x=hist_data, 
            nbinsx=30,  # Increased bins for better granularity
            name=column, 
            opacity=0.85,  # Slightly increased opacity for better visibility
            marker=dict(color=color, line=dict(color='black', width=1.5))  # High contrast borders
        ),
        row=i+1, col=1
    )
    
    # KDE trend line
    kde = gaussian_kde(hist_data)
    x_grid = np.linspace(hist_data.min(), hist_data.max(), 1000)  # Grid for the KDE curve
    kde_values = kde(x_grid)
    
    # Trend line trace (KDE curve)
    fig.add_trace(
        go.Scatter(
            x=x_grid, 
            y=kde_values * (hist_data.count() * np.diff(np.histogram(hist_data, bins=30)[1])[0]),  # Scale KDE to match histogram
            mode='lines',
            name=f'{column} Trend Line',
            line=dict(color=color, width=2, dash='dash')  # Dashed line for trend
        ),
        row=i+1, col=1
    )

# Updating layout for better visualization
fig.update_layout(
    height=350 * len(numerical_columns),  # Increased height for better spacing
    width=1000,  # Slightly wider for better readability
    showlegend=False, 
    title_text="Distributions of Numerical Features with Trend Lines", 
    title_font=dict(size=22, family="Arial", color="black"),
    template="plotly",  # Set to a more colorblind-friendly template
    paper_bgcolor="#F5F5F5",  # Light grey background for reduced glare
    plot_bgcolor="#EAEAEA",  # Slightly darker grey for plot area
    margin=dict(l=60, r=60, t=90, b=60),
    bargap=0.2  # Added gap between bars for better differentiation
)

fig.show()

* You may add as many sections as you want, as long as it supports your project workflow.
* All notebook's cells should be run top-down (you can't create a dynamic wherein a given point you need to go back to a previous cell to execute some task, like go back to a previous cell and refresh a variable content)

---

# Push files to Repo

* In cases where you don't need to push files to Repo, you may replace this section with "Conclusions and Next Steps" and state your conclusions and next steps.

In [ ]:
import os
try:
  # create your folder here
  # os.makedirs(name='')
except Exception as e:
  print(e)
